lib/ansible/galaxy/collection/__init__.py

```diff
@@ -504,8 +504,20 @@ def build_collection(u_collection_path, u_output_path, force):
     return collection_output
 
 
+def process_virtual_requirements(requirement, resolved_requirements, artifacts_manager):
+    # type: (Requirement, t.Iterable[Requirement], ConcreteArtifactsManager) -> None
+    requirement.validate_version()
+
+    if requirement.is_virtual:
+        for dep_name, dep_req in artifacts_manager.get_direct_collection_dependencies(requirement).items():
+            dep = Requirement.from_requirement_dict({'name': dep_name, 'version': dep_req}, artifacts_manager)
+            process_virtual_requirements(dep, resolved_requirements, artifacts_manager)
```

Comment:
I've got a bad feeling about the recursion here and a dependency resolution attempt outside the dependency resolver... Pretty sure the resolution should be scoped to the resolver.

Reply:
Okay, I actually went back and forth about that. My first draft was contained in the resolver, but I was modifying the preferred candidates during resolution, which seemed like it was going to be problematic. We calculate the preferred candidates outside of the dep resolver, but perhaps I can move that inside as a first step...

Reply:
I've refactored so it occurs when the dependency resolver is instantiated and put a safeguard against recursive dependencies. Since things external and internal to the dependency resolver needed this info, and `_resolve_depenency_map` is the only place that has access to the provider, it's now responsible for 1. passing the parsed requirements back into `resolve()`
and 2. handling the unsatisfied logic.

I'd like to rewrite the unsatisfied logic, but I will leave that for a follow up PR that won't be backported.